modules/sdk/src/main/scala/org/tessellation/sdk/http/p2p/middlewares/PeerAuthMiddleware.scala

```diff
@@ -0,0 +1,217 @@
+package org.tessellation.sdk.http.p2p.middleware
+
+import java.security.{PrivateKey, PublicKey}
+
+import cats.Applicative
+import cats.data.{Kleisli, OptionT}
+import cats.effect.{Async, Ref, Resource}
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+
+import org.tessellation.schema.cluster.TokenValid
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.sdk.domain.cluster.services.Session
+import org.tessellation.sdk.domain.cluster.storage.SessionStorage
+import org.tessellation.sdk.http.p2p.headers.{`X-Id`, `X-Session-Token`}
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.signature.Signing
+
+import com.comcast.ip4s.{Host, Port}
+import fs2.{Chunk, Stream}
+import org.http4s.Status.Unauthorized
+import org.http4s._
+import org.http4s.client.Client
+import pl.abankowski.httpsigner.http4s._
+import pl.abankowski.httpsigner.signature.generic.{GenericGenerator, GenericVerifier}
+import pl.abankowski.httpsigner.signature.{Generator, Verifier}
+import pl.abankowski.httpsigner.{HttpCryptoConfig, SignatureValid}
+
+object PeerAuthMiddleware {
+
+  private def unauthorized[F[_]] = Response[F](status = Unauthorized)
+
+  def responseSignerMiddleware[F[_]: Async: SecurityProvider](
+    privateKey: PrivateKey,
+    sessionStorage: SessionStorage[F]
+  )(http: HttpRoutes[F]): HttpRoutes[F] =
+    Kleisli { req: Request[F] =>
+      for {
+        res <- http(req)
+        headerToken <- sessionStorage.getToken.map(_.map(t => `X-Session-Token`(t))).attemptT.toOption
+        newHeaders = headerToken.fold(res.headers)(h => res.headers.put(h))
+        resWithHeader = res.copy(headers = newHeaders)
+        signedResponse <- new Http4sResponseSigner[F](getGenerator(privateKey), new TessellationHttpCryptoConfig {})
+          .sign(resWithHeader)
+          .attemptT
+          .toOption
+      } yield signedResponse
+    }
+
+  def responseTokenVerifierMiddleware[F[_]: Async](
+    client: Client[F],
+    session: Session[F]
+  ): Client[F] =
+    Client { (req: Request[F]) =>
+      client.run(req).flatMap { response =>
+        Resource.liftK[F] {
+          val host = req.uri.authority.map(_.host.value).flatMap(Host.fromString)
+          val port = req.uri.authority.flatMap(_.port).flatMap(Port.fromInt)
+          val token = response.headers
+            .get[`X-Session-Token`]
+            .map(_.token)
+
+          Applicative[Option]
+            .map2(host, port) { (_host, _port) =>
```

Comment:
If you already have `cats.syntax.applicative._`

Then you can just:

```scala
(host, port).map2(...)
```

or `mapN`

Reply:
```
error| value map2 is not a member of (Option[com.comcast.ip4s.Host], Option[com.comcast.ip4s.Port])
```
```
error| value mapN is not a member of (Option[com.comcast.ip4s.Host], Option[com.comcast.ip4s.Port])
```

Reply:
Try `mapN`

Reply:
ok, it was a matter of a proper import